classes/user/Collector.php

```diff
@@ -458,19 +458,29 @@ protected function buildUserGroupFilter(Builder $query): self
             ->when($this->contextIds !== null, fn ($subQuery) => $subQuery->whereIn('ug.context_id', $this->contextIds));
 
         $joinSub = clone $subQuery;
+
         $query
             ->addWhereExistsQuery(
                 $subQuery->when(
                     $this->excludeRoles != null,
-                    // This aggregates a column role_count, which holds an information whether user holds specified role
-                    fn (Builder $subQuery) => $subQuery->leftJoinSub(
-                        $joinSub
-                            ->select('uug.user_id')
-                            ->whereIn('ug.role_id', $this->excludeRoles),
-                        'agr',
-                        'uug.user_id',
-                        'agr.user_id'
-                    )->whereNull('agr.user_id')
+                    fn (Builder $subQuery) =>
+                        DB::connection()->getDoctrineConnection()->getDatabasePlatform() instanceof \Doctrine\DBAL\Platforms\MariaDBPlatform
+                        ?
+                        $subQuery->whereNotIn(
+                            'u.user_id',
+                            $joinSub
+                                ->select('uug.user_id')
+                                ->whereIn('ug.role_id', $this->excludeRoles)
+                        )
+                        :
+                        $subQuery->leftJoinSub(
+                            $joinSub
+                                ->select('uug.user_id')
+                                ->whereIn('ug.role_id', $this->excludeRoles),
+                            'agr',
+                            'uug.user_id',
+                            'agr.user_id'
+                        )->whereNull('agr.user_id')
```

Comment:
Hmm, as far as I can see, the cloned query `$joinSub`, isn't making explicit use of the outer query `$subQuery`, it's using just the `user_id` field, which is available at the main table.

So, I think we can move the `excludeRoles` stuff to the outer scope (`$query`), something like: `$query->notExists(query to select the excluded roles)`.

Reply:
Hmmm... Yes...
So instead of having:
```
select u.* from users as u
where u.disabled = 0 and exists 
(
select * from user_user_groups as uug
inner join user_groups as ug on uug.user_group_id = ug.user_group_id 
where uug.user_id = u.user_id and ug.context_id in (1) and u.user_id not IN 
(
select uug.user_id from user_user_groups as uug
inner join user_groups as ug on uug.user_group_id = ug.user_group_id 
where uug.user_id = u.user_id and ug.context_id in (1) and ug.role_id in (4096)
)
) 
order by u.user_id asc;
```

it would be:
```
select "u".* from "users" as "u" 
where "u"."disabled" = 0 and exists 
(
select * from "user_user_groups" as "uug" 
inner join "user_groups" as "ug" on "uug"."user_group_id" = "ug"."user_group_id" 
where "uug"."user_id" = "u"."user_id" and "ug"."context_id" in (1)
) 
and not exists 
(
select * from "user_user_groups" as "uug" 
inner join "user_groups" as "ug" on "uug"."user_group_id" = "ug"."user_group_id" 
where "uug"."user_id" = "u"."user_id" and "ug"."context_id" in (1) and "ug"."role_id" in (4096)
) 
order by "u"."user_id" asc
```

Which seems to be better...